### PR TITLE
🐛 Speed up ignoring terminating Pods when draining unreachable Nodes

### LINKED ISSUE #10705: Delayed MHC replacement of unreachable nodes
### What steps did you take and what happened?

- Create a CAPA cluster with at least one machine/node
- Apply a machinehealthcheck that attempts to remediate machines when nodes stop reporting status
```
spec:
  maxUnhealthy: 2
  unhealthyConditions:
  - status: Unknown
    timeout: 8m0s
    type: Ready
```
- Stop the underlying EC2 instance in AWS
- The machine is marked as unhealthy after 8 minutes, but the drain takes a long time

### What did you expect to happen?

Once the MHC determines the node is unhealthy and attempts to replace it, it should skip waiting for graceful pod termination and just delete the pods as soon as possible.

### Cluster API version

1.17.1

### Kubernetes version

v1.27.13+e709aa5

### Anything else you would like to add?

When a node is unreachable, the drain still considers the pod grace periods but skips waiting for pods to delete after 5 minutes: https://github.com/kubernetes-sigs/cluster-api/blob/0539a298e983c6dfed9771178736c345b696cc64/internal/controllers/machine/machine_controller.go#L674

We should align the behavior with OpenShift MAPI: https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/drain_controller.go#L164-L171

### Label(s) to be applied

/kind bug
/area machine

### PR DESCRIPTION
**Pull request by typeid**
**What this PR does / why we need it**:

This PR reduces the time we wait for pod deletion in the case of unreachable nodes: we currently wait for the deletionTimestamp to be 5 minutes old. The goal is to allow MHC to replace unreachable nodes faster.

This PR aligns the behavior of CAPI with OpenShift MAPI: https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/drain_controller.go#L164-L171 

This PR is a result from a discussion [here](https://github.com/kubernetes-sigs/cluster-api/pull/10662#issuecomment-2139805627).

**Which issue(s) this PR fixes** 
Fixes #10705

Area example:
/area machine

### CODE REVIEW

**Review by vincepri — COMMENTED** [review]
/lgtm

**Comment by vincepri** [comment]
It would be good though if we could add an e2e testing this behavior and expectations

**Comment by typeid** [comment]
/retest

> It would be good though if we could add an e2e testing this behavior and expectations

@vincepri I can definitely take a look at doing that. It's going to be my first shot at adding an E2E test here, so no promises. I'll check if there's something I can re-use, but the node being unreachable doesn't seem something we already have a base for.

**Comment by enxebre** [comment]
>This PR reduces the time we wait for pod deletion in the case of unhealthy nodes:

Just to clarify as this topic gets fuzzy every time we need to get back to it after a while. I would replace "unhealthy" (ambiguous) with "Unreachable" (ready = Unknown), which is the only semantic where we set this timeout because we assume those pods as gone. This PR is irrelevant for other "Unhealthy" criteria.

**Comment by typeid** [comment]
@enxebre / @vincepri The E2E doesn't seem trivial. I would skip adding the E2E for now if that's okay. Shall I create a follow-up issue to add functionality in our test framework to render nodes unreachable? This would allow us to improve our coverage in other areas as well. 

Adding a unit test here would also not add any coverage of value IMO and require more code changes.

**Comment by enxebre** [comment]
I'm good with that, I'll defer to Vince to cancel the hold as he considers
/approve
/hold

File: internal/controllers/machine/machine_controller.go
Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:683]:
```go
		// The grace period generally won't change anything here as
		// the kubelet should not be reachable, but it doesn't hurt setting it.
```

I'm a bit confused. If it doesnt change anything, why are we setting it?

Comment (typeid) [reply internal/controllers/machine/machine_controller.go:683]: The main reason I added this is that it aligns with [OpenShift MAPI](https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/drain_controller.go#L164-L171). I was not able to think of an edge case this covers, but it also has no negative effect.

As mentioned though, we could also change the value to `0` for it to actually have the effect of pods being force deleted by the kube-apiserver instead of deletion being skipped. Although I was not able to find the behavior of `0` documented, so that might need some additional thorough digging in kube-apiserver. 

@enxebre any thoughts on this? I'm happy to reach out to the MAPI folks to see if we're missing some reason for the value to be set to `1` there.

Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:683]: It would be good to have a reason to set this. Otherwise we have no way to reason about changing or not changing this value going forward.

Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:683]: Btw nothing against OpenShift MAPI, but I don't want to make changes to CAPI just to align to OpenShift.

Comment (enxebre) [reply internal/controllers/machine/machine_controller.go:683]: The kas sets the deletion timestamp to a future date compute the grace period [1].
So effectively we wait SkipWaitForDeleteTimeoutSeconds + GracePeriodSeconds, hence setting it to 1 makes sense to keep it short. (0 would also work but I'd prefer 1 for alignment).

[1] https://github.com/kubernetes/kubernetes/blob/5bf1e95541d90e37f6c6637b5b45d8783e7907aa/staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go#L150

This is where we added to mapi https://github.com/openshift/machine-api-operator/pull/780
This is the driving bug https://bugzilla.redhat.com/show_bug.cgi?id=1905709
These are related features
https://github.com/kubernetes/kubernetes/pull/94381
https://github.com/kubernetes/kubernetes/pull/91342

So maybe just update the comment to reflect the above. cc @sbueringer

File: internal/controllers/machine/machine_controller.go
Comment (enxebre) [reply internal/controllers/machine/machine_controller.go:684]: your assumption is correct. But not because we don't honour PDBs. We still make the evict API request which account for pdbs. But because of https://github.com/kubernetes/kubernetes/pull/94381 and https://github.com/kubernetes/kubernetes/pull/91342 the request legitimately succeeds as the unhealthy pods do not decrease the disruption budget anymore. Before those fixes drain might block forever while deleting an unreachable node even though the pods and nodes were doomed.

Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:684]: Pods on unreachable nodes (ie ready: unknown) are automatically marked as unhealthy? (I guess by setting the pod ready condition to unknown)

Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:684]: I couldn't find it explicitly called out in the docs.

But looks like at the same time when the Ready condition of the Node is flagged as Unknown the Ready condition of the Pod is set to False

```
# Node:
  - lastHeartbeatTime: "2024-06-14T09:25:35Z"
    lastTransitionTime: "2024-06-14T09:27:48Z"
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: Ready

# Pod
  - lastProbeTime: null
    lastTransitionTime: "2024-06-14T09:27:48Z"
    status: "False"
    type: Ready
```

Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:684]: For reference. This seems to be the code that marks Pods as not ready when the Node is not ready: https://github.com/kubernetes/kubernetes/blob/6d0ac8c561a7ac66c21e4ee7bd1976c2ecedbf32/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1151

Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:684]: Back to the original point

> your assumption is correct. But not because we don't honour PDBs.

Thx, I took another look at the drain helper and eviction and yeah this is obviously true :joy:

**Comment by sbueringer** [comment]
/lgtm

Thx everyone for the explanations!

/hold
Similar to the other PR, I asked some other folks for their opinion. I'll get back to you ASAP (I expect early next week)

**Review by vincepri — APPROVED** [review]
/hold cancel
/lgtm

**Comment by sbueringer** [comment]
/hold
Similar to the other PR, I asked some other folks for their opinion. I'll get back to you ASAP (I expect early next week)

If that's okay?

**Comment by vincepri** [comment]
Sorry I was under the impression per Alberto's comment that we were waiting for my review 😄

**Comment by sbueringer** [comment]
Thx everyone!

/hold cancel

File: internal/controllers/machine/machine_controller.go
Comment (enxebre) [reply internal/controllers/machine/machine_controller.go:684]: also for record the node life cycle controller automatically starts evicting all the pods when a kubelet is unreachable for longer than N https://kubernetes.io/docs/concepts/architecture/nodes/#node-controller

**Pull request merged**
